nexus/src/app/mod.rs

```diff
@@ -210,9 +211,19 @@ impl Nexus {
                 "component" => "DpdClient"
             )),
         };
-        let dpd_address = config.pkg.dendrite.address;
-        let dpd_host = dpd_address.ip().to_string();
-        let dpd_port = dpd_address.port();
+        let (dpd_host, dpd_port) = if let Some(dpd_address) =
+            &config.pkg.dendrite.address
+        {
+            (dpd_address.ip().to_string(), dpd_address.port())
+        } else {
+            let addr = resolver
+                .lock()
+                .await
+                .lookup_socket_v6(ServiceName::Dendrite)
+                .await
+                .map_err(|e| format!("Cannot access Dendrite address: {e}"))?;
```

Comment:
So, while building this, I figured I'd mention - it's a little weird that we only have *one* `dpd_client` -- there are multiple scrimlets, so should we have a `Vec` of `dpd_client`s?

Reply:
Probably, yes. Which one(s) need to be used depends on the operation, but yeah, we'll need to be able to talk to both. I don't know how to distinguish them at the level of DNS, though.

Reply:
Pasting this here, I screwed up where I typed this: I'm pretty sure that when creating the service plan with multiple scrimlets, after this PR there will be multiple dendrite addresses for the Dendrite DNS record. This does mean that:

- RSS will have to wait for two scrimlets to appear as peers
- we should be creating dpd clients dynamically based on the DNS instead

Reply:
To be clear - instead of once and storing it in the Nexus object

Reply:
I filed https://github.com/oxidecomputer/omicron/issues/2760 - this should be possible by iterating over the results returned from the resolver: https://github.com/oxidecomputer/omicron/blob/e031c290bd09e1a5114427b5172256b73d697726/internal-dns/src/resolver.rs#L111-L114

Reply:
I might be missing something, but we really need to treat every `dpd` client as distinct. If I remember the DNS parlance correctly, they are each different services, with exactly one backend. Is that what's happening here?

Reply:
@bnaecker this PR should be generating separate AAAA records for each dendrite zone, but you're right, I believe this generates **one** SRV record to represent both dendrite services, even though they're distinct.

Do we have any way of distinguishing "switch 1" from "switch 2" currently? I could label them with SRV records that distinguish the dendrite instances from each other explicitly, but I want some reliable way of distinguishing them

Reply:
> Do we have any way of distinguishing "switch 1" from "switch 2" currently? I could label them with SRV records that distinguish the dendrite instances from each other explicitly, but I want some reliable way of distinguishing them

You could (probably?) do this via MGS the same way we've talked about identifying gimlets: if there was a way for the scrimlet to self-report its switch's identifying info (serial number/board/revision), you could compare that against the identifying info MGS reports for "switch 1" vs "switch 2".

Reply:
I agree the best answer is something like what's laid out in RFD 360 like you're saying: model / revision / serial as a unique ID. Unfortunately, I'm not sure how we get that data currently from host software on the Scrimlet. We've implemented a way to fetch the MAC address range programmed into the FRUID EEPROM, but not those other fields.

Reply:
So, if this is definitely the data we want to use for identifying the DNS records, then we'll probably need to add a message to the `transceiver-controller` crate for fetching it. See [this issue](https://github.com/oxidecomputer/transceiver-control/issues/73) for tracking.

From talking more with Robert, it sounds like we probably want two different schemes in DNS for services like `dpd`:

### `<serial>.<type>.<az>`

- Serial number (from this hypothetical new message above),
- The type of machine the program runs on: `sled`, `switch` or `psc` at this point,
- The AZ, either a number or UUID, not sure yet.

### `<loc>.<rack>.<az>`

- A rack location, e.g., `sled23`, `switch0` or `psc0`,
- A rack number or UUID, not sure yet.
- An AZ number or UUID, not sure yet.


It seems like populating the serial-number-based DNS records would require a request to `dpd` (at this point listening on `::1`) and out to the SP over the management network. From John's comment, it sounds like the location-based name could come from MGS, but I'm definitely guessing.

Reply:
I still think this is worth fixing, but I filed https://github.com/oxidecomputer/omicron/issues/2760 as a separate issue, where we can follow-up.